subprojects/core/src/main/java/org/gradle/normalization/internal/DefaultRuntimeClasspathNormalization.java

```diff
@@ -110,14 +74,49 @@ public void ignoreManifest() {
 
         @Override
         public void ignoreAttribute(String name) {
-            checkNotEvaluated();
-            attributeIgnoresBuilder.add(name.toLowerCase(Locale.ROOT));
+            manifestAttributeResourceFilter.ignore(name.toLowerCase(Locale.ROOT));
         }
 
         @Override
         public void ignoreProperty(String name) {
+            manifestPropertyResourceFilter.ignore(name);
+        }
+    }
+
+    static <T> EvaluatableFilter<T> filter(Function<ImmutableSet<String>, T> initializer, T emptyValue) {
+        return new EvaluatableFilter<>(initializer, emptyValue);
+    }
+
+    private static class EvaluatableFilter<T> {
+        private T value;
+        private final Supplier<T> valueSupplier;
+        protected final ImmutableSet.Builder<String> builder;
+
+        EvaluatableFilter(Function<ImmutableSet<String>, T> initializer, T emptyValue) {
+            this.builder = ImmutableSet.builder();
+            // if there are configured ignores, use the initializer to create the value, otherwise return emptyValue
+            this.valueSupplier = () -> Optional.of(builder.build())
+                                                .filter(ignores -> !ignores.isEmpty())
+                                                .map(initializer)
+                                                .orElse(emptyValue);
+        }
+
+        T evaluate() {
+            if (value == null) {
+                value = valueSupplier.get();
+            }
+            return value;
+        }
```

Comment:
Let's use `public` for methods even in `private` inner classes. That's what we do elsewhere, too. The rationale is that if you change the classes visibility to, say, `public`, these methods should also be visible with it.

Reply:
Fair enough - done.